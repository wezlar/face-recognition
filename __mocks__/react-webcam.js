// todo not sure if this is still used. May need deleting
import React from 'react';

class Webcam extends React.Component {

  render() {
    return null;
  }
}

export default Webcam;