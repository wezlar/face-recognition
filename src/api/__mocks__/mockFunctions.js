export const mockFindBestMatch = {
  _label: 'Rio',
};

export const mockGetFullFaceDescription = [
  {
    descriptor: [
      [-0.08441604673862457,0.052526094019412994,0.038848020136356354,-0.08337470144033432,-0.04573802649974823,-0.04051077365875244,-0.03232140094041824,-0.007713532540947199,0.2459832727909088,0.00047820876352488995,0.13936340808868408,0.015116522088646889,-0.22676977515220642,-0.058322835713624954,0.0004984699189662933,0.08946414291858673,-0.16845925152301788,-0.08100305497646332,-0.12426283210515976,-0.09014890342950821,-0.035736262798309326,0.0744253620505333,0.1325514018535614,0.011960530653595924,-0.07576292753219604,-0.33382588624954224,-0.08692063391208649,-0.21836167573928833,0.11622339487075806,-0.20780114829540253,-0.009222961962223053,-0.042087215930223465,-0.18633879721164703,-0.13667942583560944,0.029456188902258873,-0.009402288123965263,-0.07245419174432755,-0.13483040034770966,0.246804341673851,0.07715440541505814,-0.11998458951711655,0.05225367099046707,-0.020490575581789017,0.23290377855300903,0.16812576353549957,0.01925789751112461,0.010702788829803467,-0.06880012154579163,0.14114898443222046,-0.2781291902065277,0.10215630382299423,0.1268511712551117,0.11517145484685898,0.05802106112241745,0.13286517560482025,-0.1947886347770691,0.07148679345846176,0.15228517353534698,-0.2996051013469696,0.20219314098358154,0.04039623588323593,-0.014870991930365562,-0.05420020967721939,0.03607364743947983,0.19460201263427734,0.13236680626869202,-0.1492900550365448,-0.15047655999660492,0.022033775225281715,-0.13819371163845062,0.08385756611824036,0.16245631873607635,-0.09493327140808105,-0.1958199292421341,-0.19934304058551788,0.06532489508390427,0.38936060667037964,0.09247858077287674,-0.1653134822845459,-0.002613373100757599,-0.14572389423847198,-0.06388683617115021,-0.028549285605549812,0.09196682274341583,-0.13314640522003174,-0.037275828421115875,-0.025153793394565582,-0.03427465632557869,0.08105505257844925,0.04603111371397972,-0.08276700973510742,0.24077735841274261,-0.05688551068305969,0.06049064174294472,-0.004596324637532234,0.02712996117770672,-0.14758889377117157,-0.08222685754299164,0.02918913960456848,-0.028991293162107468,0.024095753207802773,-0.18331533670425415,-0.04099935293197632,0.03288392722606659,-0.19606611132621765,0.11679170280694962,-0.03816153481602669,-0.046753738075494766,0.0016085415845736861,-0.004431181587278843,-0.042677924036979675,-0.005630177445709705,0.17890341579914093,-0.369128555059433,0.28752145171165466,0.2323327213525772,0.01609717309474945,0.15800492465496063,-0.008518354967236519,0.02593380957841873,0.0021005170419812202,0.07626444101333618,-0.07457248866558075,-0.040228940546512604,-0.028733637183904648,-0.021051086485385895,0.02640320546925068,-0.004543062299489975]
    ],
    detection: {
      box: {
        height: 156.0831506280933,
        width: 293.60938969101437,
        _x: 293.60938969101437,
        _y: 88.34867537841333,
      },
    },
  },
];
