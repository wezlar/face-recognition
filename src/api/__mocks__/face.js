import { mockFindBestMatch, mockGetFullFaceDescription } from './mockFunctions'; 

module.exports = {
  loadModels: jest.fn(),
  getFullFaceDescription: jest.fn().mockResolvedValue(mockGetFullFaceDescription),
  createMatcher: jest.fn().mockResolvedValue({
    findBestMatch: jest.fn().mockReturnValue(mockFindBestMatch)
  }),
};