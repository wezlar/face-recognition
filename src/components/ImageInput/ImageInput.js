import React, { Component } from 'react';
import { loadModels, getFullFaceDescription, createMatcher } from '../../api/face';
import DrawBox from '../DrawBox/DrawBox';

import { JSON_PROFILE, INIT_STATE } from '../../constants';

class ImageInput extends Component {
  constructor(props) {
    super(props);
    this.state = { ...INIT_STATE };
  }

  componentWillMount = async () => {
    await loadModels();
    this.setState({ faceMatcher: await createMatcher(JSON_PROFILE) });
    await this.handleImage(this.state.imageURL);
  };

  handleImage = async (image = this.state.imageURL) => {
    await getFullFaceDescription(image).then(fullDesc => {
      console.log({fullDesc});
      if (!!fullDesc) {
        this.setState({
          fullDesc,
          detections: fullDesc.map(fd => fd.detection),
          descriptors: fullDesc.map(fd => fd.descriptor),
        });
      }
    });
    
    if (!!this.state.descriptors && !!this.state.faceMatcher) {
      let match = await this.state.descriptors.map(descriptor =>
        this.state.faceMatcher.findBestMatch(descriptor)
      );
      this.setState({ match });
    }
  };

  handleFileChange = async event => {
    this.resetState();
    await this.setState({
      imageURL: URL.createObjectURL(event.target.files[0]),
      loading: true,
    });
    this.handleImage();
  };

  resetState = () => {
    this.setState({ ...INIT_STATE });
  };

  drawDescriptorsInput = () => {
    const { descriptors } = this.state;
    if (!descriptors) {
      return null;
    }
  
    return (<input
      id='myFileDescriptors'
      type='text'
      defaultValue={descriptors.join(', ')} />);
  };

  render() {
    const { imageURL, detections, descriptors, match } = this.state;

    return (
      <div>
        <input
          id='myFileUpload'
          type='file'
          onChange={this.handleFileChange}
          accept='.jpg, .jpeg, .png' />
        {this.drawDescriptorsInput()}
        <div style={{ position: 'relative' }}>
          <div style={{ position: 'absolute' }}>
            <img src={imageURL} alt='imageURL' />
          </div>
          <DrawBox detections={detections} match={match} />
        </div>
      </div>
    );
  }
}

export default ImageInput;