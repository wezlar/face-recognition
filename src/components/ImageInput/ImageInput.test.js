// __tests__/fetch.js
import React from 'react';
import {
  render,
  cleanup,
  waitForElement,
} from '@testing-library/react';
// this adds custom jest matchers from jest-dom
import ImageInput from './ImageInput';
import { mockFindBestMatch } from '../../api/__mocks__/mockFunctions';

jest.mock('../../api/face');

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup);
describe(`Test ImageInput`, () => {
  test(`Test snapshot`, async () => {
    // Arrange
    const { container, debug, getByTestId } = render(
      <ImageInput />,
    );

    // snapshots work great with regular DOM nodes!
    expect(container.firstChild).toMatchSnapshot();

    const testing = await waitForElement(() => getByTestId('draw-box'));
    expect(testing).toMatchSnapshot();
    expect(getByTestId('draw-name-box').textContent).toBe(mockFindBestMatch._label);
  });
});