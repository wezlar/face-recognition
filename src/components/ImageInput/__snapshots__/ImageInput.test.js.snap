// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ImageInput Test snapshot 1`] = `
<div>
  <input
    accept=".jpg, .jpeg, .png"
    id="myFileUpload"
    type="file"
  />
  <div
    style="position: relative;"
  >
    <div
      style="position: absolute;"
    >
      <img
        alt="imageURL"
        src="file-stub"
      />
    </div>
  </div>
</div>
`;

exports[`Test ImageInput Test snapshot 2`] = `
<div
  data-testid="draw-box"
  style="position: absolute; border: solid; border-color: blue; height: 156.0831506280933px; width: 293.60938969101437px; transform: translate(293.60938969101437px, 88.34867537841333px);"
>
  <p
    data-testid="draw-name-box"
    style="background-color: blue; border: solid; border-color: blue; margin-top: 0px; color: rgb(255, 255, 255); width: 293.60938969101437px; transform: translate(-3px, 156.0831506280933px);"
  >
    Rio
  </p>
</div>
`;
